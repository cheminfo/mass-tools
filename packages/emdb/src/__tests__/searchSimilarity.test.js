'use strict';

const DBManager = require('..');

describe('test searchSimilarity', () => {
  it('should find one result with bad distribution', async () => {
    let dbManager = new DBManager();
    await dbManager.loadNeutralTest({ maxC: 10 });
    dbManager.setExperimentalSpectrum({ x: [41, 121], y: [1, 1] });
    let results = await dbManager.searchSimilarity({
      ionizations: 'H+,(H+)2,(H+)3', // useless because the test database has already ionizations
      filter: {},
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 2.5,
        },
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });

    expect(results.test).toHaveLength(2);

    expect(results.test).toMatchSnapshot();
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.886, 3);
    expect(results.test[1].ms.similarity.value).toBeCloseTo(0.886, 3);
  });

  it('should find one result with callback', async () => {
    let dbManager = new DBManager();
    await dbManager.loadNeutralTest({ maxC: 10 });
    dbManager.setExperimentalSpectrum({ x: [41, 121], y: [1, 1] });
    let results = await dbManager.searchSimilarity({
      ionizations: 'H+,(H+)2,(H+)3', // useless because the test database has already ionizations
      filter: {
        callback: (entry) => entry.atoms.C === 10,
      },
    });

    expect(results.test).toHaveLength(2);
  });

  it('should find one result with wrong callback', async () => {
    let dbManager = new DBManager();
    await dbManager.loadNeutralTest({ maxC: 10 });
    dbManager.setExperimentalSpectrum({ x: [41, 121], y: [1, 1] });
    let results = await dbManager.searchSimilarity({
      ionizations: 'H+,(H+)2,(H+)3', // useless because the test database has already ionizations
      filter: {
        callback: (entry) => entry.atoms.C === 9,
      },
    });

    expect(results.test).toHaveLength(0);
  });

  it('should find one result with perfect match (small zone)', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {},
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 0.5,
        },
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    //   expect(results.test).toMatchSnapshot();
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with bad distribution, small zone, small width', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 0.5,
        },
        widthBottom: 0.1,
        widthTop: 0.1,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });

    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find no result because of filter', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1,
      },
    });
    expect(results.test).toHaveLength(0);
  });

  it('should find one result with bad bad distribution, large window', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 2.5,
        },
        widthBottom: 0.1,
        widthTop: 0.1,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.895, 2);
  });

  it('should find one result with bad bad distribution, large window huge width', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 2.5,
        },
        widthBottom: 5,
        widthTop: 5,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with overlap', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 2.5,
        },
        common: 'first', // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with good distribution', async () => {
    let dbManager = new DBManager();
    await dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 0.11] });
    let results = await dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.995, 2);
  });
});
