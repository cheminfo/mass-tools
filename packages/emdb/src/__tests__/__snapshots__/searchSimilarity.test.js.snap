// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test searchSimilarity > should find one result with bad distribution 1`] = `
[
  {
    "atoms": {
      "C": 10,
    },
    "charge": 0,
    "em": 120,
    "ionization": {
      "atoms": {
        "H": 3,
      },
      "charge": 3,
      "em": 3.02347509669,
      "mf": "(H+)3",
    },
    "mf": "C10",
    "ms": {
      "charge": 3,
      "em": 41.00727645232093,
      "ionization": "(H+)3",
      "similarity": {
        "difference": [
          [
            41.00727645232093,
            41.341731161256995,
            41.676186546352326,
          ],
          [
            0.011646827156351125,
            0.09741601219811298,
            0.004759776692041084,
          ],
        ],
        "experimental": [
          [
            41,
          ],
          [
            1,
          ],
        ],
        "experimentalInfo": {
          "from": 40.840609785654266,
          "max": 0.5,
          "min": 0.5,
          "sum": 0.5,
          "to": 41.840609785654266,
        },
        "quantity": 0.5,
        "theoretical": [
          [
            41.00727645232093,
            41.341731161256995,
            41.676186546352326,
          ],
          [
            0.8978242111098459,
            0.09741601219811298,
            0.004759776692041084,
          ],
        ],
        "thereoticalInfo": {
          "max": 0.8976979854295873,
          "min": 0.004759107511990694,
          "sum": 0.9998594093602103,
        },
        "value": 0.8861773839534948,
        "width": {
          "bottom": 0.05,
          "top": 0.01,
        },
      },
    },
    "mw": 120.10735896735248,
    "parts": [
      "C10",
    ],
    "unsaturation": 11,
  },
  {
    "atoms": {
      "C": 10,
    },
    "charge": 0,
    "em": 120,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C10",
    "ms": {
      "charge": 1,
      "em": 121.00727645232094,
      "ionization": "H+",
      "similarity": {
        "difference": [
          [
            121.00727645232094,
            122.01063439121715,
            123.01399301090913,
          ],
          [
            0.01185238253244636,
            0.09723174483189419,
            0.004738488682519898,
          ],
        ],
        "experimental": [
          [
            121,
          ],
          [
            1,
          ],
        ],
        "experimentalInfo": {
          "from": 120.50727645232094,
          "max": 0.5,
          "min": 0.5,
          "sum": 0.5,
          "to": 123.50727645232094,
        },
        "quantity": 0.5,
        "theoretical": [
          [
            121.00727645232094,
            122.01063439121715,
            123.01399301090913,
          ],
          [
            0.8980297664855859,
            0.09723174483189419,
            0.004738488682519898,
          ],
        ],
        "thereoticalInfo": {
          "max": 0.8979044915878657,
          "min": 0.004737827664692645,
          "sum": 0.9998605002836259,
        },
        "value": 0.8861773839531395,
        "width": {
          "bottom": 0.05,
          "top": 0.01,
        },
      },
    },
    "mw": 120.10735896735248,
    "parts": [
      "C10",
    ],
    "unsaturation": 11,
  },
]
`;
